omniledger/service/service.go

```diff
@@ -366,44 +379,91 @@ func (s *Service) createQueueWorker(scID skipchain.SkipBlockID) (chan Transactio
 // We use the omniledger as a receiver (as is done in the identity service),
 // so we can access e.g. the collectionDBs of the service.
 func (s *Service) verifySkipBlock(newID []byte, newSB *skipchain.SkipBlock) bool {
-	_, dataI, err := network.Unmarshal(newSB.Data, cothority.Suite)
-	data, ok := dataI.(*Data)
+	_, headerI, err := network.Unmarshal(newSB.Data, cothority.Suite)
+	header, ok := headerI.(*DataHeader)
+	if err != nil || !ok {
+		log.Errorf("couldn't unmarshal header")
+		return false
+	}
+	_, bodyI, err := network.Unmarshal(newSB.Payload, cothority.Suite)
+	body, ok := bodyI.(*DataBody)
 	if err != nil || !ok {
-		log.Errorf("couldn't unmarshal data")
+		log.Error("couldn't unmarshal body", err, ok)
 		return false
 	}
-	txs := data.Transactions
+
+	txs := body.Transactions
+	var ctx []ClientTransaction
+	for _, t := range txs {
+		ctx = append(ctx, t.ClientTransaction)
+	}
 	cdb := s.getCollection(newSB.Hash)
-	for _, tx := range txs {
-		f, exists := s.verifiers[string(tx.Kind)]
-		if !exists || tx.Valid != f(cdb, &tx) {
-			return false
+	mtr, ots := s.createOmniledgerTransactions(cdb.coll, ctx)
+	if bytes.Compare(mtr, header.CollectionRoot) != 0 {
+		log.Lvl2(s.ServerIdentity(), "Collection root doesn't verify")
+		return false
+	}
+	hash := sha256.New()
+	for _, tx := range ots {
+		txBuf, err := protobuf.Encode(&tx)
+		if err != nil {
+			log.Lvl2(s.ServerIdentity(), "Couldn't marshal transaction")
 		}
+		hash.Write(txBuf)
+	}
+	if bytes.Compare(hash.Sum(nil), header.TransactionHash) != 0 {
+		log.Lvl2(s.ServerIdentity(), "Transaction hash doesn't verify")
+		return false
 	}
 	return true
 }
 
-// validateTransactions sets the valid-flag of the transaction according to the
-// registered OmniledgerVerifiers.
-func (s *Service) validateTransactions(cdb *collectionDB, txs []Transaction) {
-	for i := range txs {
-		f, exists := s.verifiers[string(txs[i].Kind)]
-		// If the leader does not have a verifier for this kind, it drops the
-		// transaction.
-		if !exists {
-			log.Printf("%+v", s.verifiers)
-			log.Lvl1("Leader is dropping transaction of unknown kind:", string(txs[i].Kind))
-			txs = append(txs[:i], txs[i+1:]...)
-			continue
+// createOmniledgerTransactions goes through all ClientTransactions and creates
+// the appropriate StateChanges and sets the valid flag.
+func (s *Service) createOmniledgerTransactions(coll collection.Collection, txs []ClientTransaction) ([]byte, []OmniledgerTransaction) {
+	cdbTemp := coll.Clone()
+	var otx []OmniledgerTransaction
+clientTransactions:
+	for i, ctx := range txs {
+		cdbI := cdbTemp.Clone()
+		ot := OmniledgerTransaction{ClientTransaction: ctx, Valid: true}
+		for _, instr := range ctx.Instructions {
+			kind, state, err := instr.GetKindState(cdbI)
+			if err != nil {
+				log.Lvl1("Couldn't get kind of instruction")
+				txs = append(txs[:i], txs[i+1:]...)
```

Comment:
I usually avoid modifying the slice that you're iterating inside the iteration because I'm never sure if the language does what you expect it to do and it's also confusing, at least to me. why do we need to do `txs = append(...)` actually? at the top of the inner for-loop we cat set `ot` to the default value, i.e. `Valid: false` and then set it to true if all the inner for-loops are ok.